### ISSUE TITLE
[New] CountryData register the total number of provinces as country level data

### ISSUE THREAD

**Issue by andrybicio** [lisphilar/covid19-sir#555]
## Summary
When importing a CountryData instance it does not store the country, so when instantiating a scenario it raises an error. 

## Codes
```Python

from covid19dh import covid19
import covsirphy as cs
import numpy as np


# Dataset preparation

x, src = covid19("Italy", level = 2)

#the following lines are to cast to integers, since when manually using as input a csv file it casts automatically to floats 
#and then it raises an error
v_cols = ['deaths', 'confirmed', 'recovered', 'deaths']
x[v_cols] = x[v_cols].fillna(0).astype(np.int64)

#creating csv files
src.to_csv("input/italy_src.csv")
x.to_csv("input/italy.csv")

#manually create a CountryData instance and specifying the country
# Create CountryData instance
country_data = cs.CountryData("input/italy.csv", country="Italy")
country_data.set_variables(
    date="date",
    confirmed="confirmed",
    recovered="recovered",
    fatal="deaths",
    province="administrative_area_level_2"
)

jhu_data = cs.JHUData.from_dataframe(country_data.cleaned())
jhu_data.countries()
```
> []

it returns an empty list

Then some of my code is missing for adding population to each province/region. I can provide if needed.
Finally:

```Python

# Scenario analysis
snl = cs.Scenario(jhu_data, population_data, country="Italy", province=None)
```

## Outputs

> ---------------------------------------------------------------------------
> SubsetNotFoundError                       Traceback (most recent call last)
> <ipython-input-116-a308761d15a1> in <module>
> ----> 1 snl = cs.Scenario(jhu_data, population_data, country="Italy", province=None)
> 
> ~/anaconda3/lib/python3.7/site-packages/covsirphy/analysis/scenario.py in __init__(self, jhu_data, population_data, country, province, tau, auto_complement)
>      34             jhu_data, population_data, country, province=province, auto_complement=auto_complement)
>      35         self.tau = self.ensure_tau(tau)
> ---> 36         self._update_range(first_date=None, last_date=None)
>      37         # Linear regression modes for prediction of ODE parameters
>      38         self._oxcgrt_data = None
> 
> ~/anaconda3/lib/python3.7/site-packages/covsirphy/analysis/scenario.py in _update_range(self, first_date, last_date)
>     102         if last_date is not None:
>     103             self._last_date = self._ensure_date_in_range(last_date)
> --> 104         super().init_records()
>     105         self._init_trackers()
>     106 
> 
> ~/anaconda3/lib/python3.7/site-packages/covsirphy/analysis/data_handler.py in init_records(self)
>      56             start_date=self._first_date, end_date=self._last_date,
>      57             population=self.population,
> ---> 58             auto_complement=self._auto_complement
>      59         )
>      60         # First/last date of the records
> 
> ~/anaconda3/lib/python3.7/site-packages/covsirphy/cleaning/jhu_data.py in records(self, country, province, start_date, end_date, population, auto_complement, **kwargs)
>     568         if auto_complement:
>     569             df, is_complemented = self.subset_complement(
> --> 570                 **subset_arg_dict, **kwargs)
>     571             if not df.empty:
>     572                 return (df, is_complemented)
> 
> ~/anaconda3/lib/python3.7/site-packages/covsirphy/cleaning/jhu_data.py in subset_complement(self, country, province, start_date, end_date, population, **kwargs)
>     513             raise SubsetNotFoundError(
>     514                 country=country, country_alias=country_alias, province=province,
> --> 515                 start_date=start_date, end_date=end_date) from None
>     516         # Complement, if necessary
>     517         self._recovery_period = self._recovery_period or self.calculate_recovery_period()
> 
> SubsetNotFoundError: Records in Italy were not found.
> 

## Environment
- CovsirPhy version: 
- Python version: 3.7.9
- Installation: pip
- System: Ubuntu 18.04 LTS

**Comment by lisphilar** [comment]
Thank you for your feedback!
`covid19("Italy", level=2)` does not have country level data, but country level data was selected with `cs.Scenario(..., country="Italy", province=None)`.
When you want to use total values at province level as country level data, please run `CountryData.set_variables()` with `province=None` (or just comment out `province="administrative_area_level_2"`).
```Python
country_data.set_variables(
    date="date",
    confirmed="confirmed",
    recovered="recovered",
    fatal="deaths",
    # Not specify province column to use total values of all provinces as country level data
    # province="administrative_area_level_2"
)
```
All outputs: https://gist.github.com/lisphilar/4b9af6122fcdb116430bbab2fda5bff0

**Comment by andrybicio** [comment]
Thank you for your reply! 
Does it mean that several csv files must be created, one for every province? Thus we will be treating each of them as several single countries?

**Comment by lisphilar** [comment]
One CSV file is enough, but two instance sets of `CountryData` and `JHUdata` should be created. The first one is with `set_varables(province=None)` and the second is with `set_variables(province="administrative_area_level_2")`.

By the way, did you tried `DataLoader.jhu()`? In your script, `covid19dh` package retrieves the dataset from "COVID-19 Data Hub" database. `DataLoader.jhu()` downloads datasets from this and produce `JHUData` instance.

**Comment by andrybicio** [comment]
I tried, but I had some issues when using a finer resolution ("administrative_area_level_3", that in Italy are districts), such as some negative values occurring, thenk I kept going on without using the dataloader. It's more straightforward, indeed. Using the two instances sets now it works, thank you!

Just a humble suggestion: it could be helpful to pass an optional parameter dictionary when importing a csv file in the "cs.CountryData()" to get its values in the proper format (if number of recovered/infected/susceptible/etc is in float it raises an error), as there in pandas.read_csv (the dtype = {'col1': type1 , 'col2': type2 ...etc }  one).

**Comment by lisphilar** [comment]
1. register total value of all provinces as country level data
As an enhancement, I will try to add `CountryData.register_total()` to register total value of all provinces as country level data with #556. With this change the codes will be as follows.
```Python
raw_df, *_ = covid19dh.covid19("Italy", level=2, verbose=False)
filename = "input/italy_raw.csv"
raw_df.to_csv(filename)
# Create CountryData instance
country_data = CountryData(filename=filename, country="Italy")
country_data.set_variables(
        date="date", confirmed="confirmed", recovered="recovered", fatal="deaths",
        province="administrative_area_level_2"
    )
# Register total value of all provinces as country level data
country_data.register_total()
# Scenario analysis at country level
snl = cs.Scenario(jhu_data, population_data, country="Italy", province=None)
# Scenario analysis at province level
snl = cs.Scenario(jhu_data, population_data, country="Italy", province="Sardegna")
```
Will it be useful for your analysis?
(Documentation is not updated at this time.)

2. data format
Thank you, but this seems a error of `country_data.cleaned()`. With #556, I will try this with `df[col] = pd.to_numeric(df[col], errors="coerce")` for value columns.

3. "administrative_area_level_3" data (common name is "city"?)
Will it be useful to prepare dataloader for this level? I do not have much time to investigate this dataset today, but, if useful, this could be discussed in a new issue.

**Comment by andrybicio** [comment]
1.  This could be indeed very useful, since from the documentation I couldn't understand what later you explained me! I thought country variable was already implemented, and could not get the error. Then I understood how to deal with datasets, but only after discussing it here. For these reasons the method "register_total()" to update also data at level country is useful for sure, to avoid confusion. Thank you!

2. Ok, I saw  the pull request and indeed that was the error I faced due to data type (expected int, got float).

3. Might be useful, but only if you have time of course! No rush! In Italy we have for instance:
- State (country level) ->  Italy
- Region -> Sardegna  (contained in administrative_area_level_2)
- Province -> Sassari (contained in administrative_area_level_3)

If I understood what CountryData.register_total() does, one could implement a similar function CountryData.register_total() but at at a regional resolution. Something like RegionData.register_total(). In other words: from _administrative_area_level_3_ resolution data, obtain the totals of _administrative_area_level_2_ using a similar function to the one you are going to implement.  "Provinces" will have now 2 layers, and not 1 any more. 
Hope these feedbacks might be helpful.

**Comment by lisphilar** [comment]
1. Pull request #556 was merged! Please use it with the latest development version 2.15.0-gamma (`pip install --upgrade "git+https://github.com/lisphilar/covid19-sir.git#egg=covsirphy"`). The next stable version is planned on 07Feb2021.

4. Documentation of dataset preparation guide
Do you have any ideas regarding updating documentation?
https://lisphilar.github.io/covid19-sir/INSTALLATION.html#use-a-local-csv-file-which-has-the-number-of-cases

Actually, the current developers, including me, create dataset classes certainly with calling `DataLoader` class for analysis.

2. This was also fixed in the merged pull request. Just for issue tracking, I created #557.

3. Three layers (level 1, 2, 3) is expected, but it is difficult to implement this feature without breaking the system of CovsirPhy at this time. We used only country and province as levels because dataset did not have city level data when started this project....
I think we can skip this problem with `CountryData` by regarding province (level 2) as country (level 1), but failed with errors.
Please refer to https://gist.github.com/lisphilar/267f334431be9af5beb27503bb3734ca

Do you have time to join our project as a developer? :)

**Comment by andrybicio** [comment]
1.  Cool thank you!
2. I would add a disclaimer that for some countries the deepest level becomes the "province" no matter wat, and the country includes both level 1 (i.e. Country) and level 2 (i.e. Regions, for Italy). Some lines of code as in the gist link can be enough. Code cells from [7] to [17].   
3. However there is a problem in the code here.  If using "Country/Region" format for the "country" parameter, I would split the parameter string as "Country"+"/"+"Region" and then upload the csv data, by first subsetting it and including the following the condition "administrative_area_level_2" == "Region" (example "administrative_area_level_2" == "Sardegna"). Since here https://gist.github.com/lisphilar/267f334431be9af5beb27503bb3734ca you are including districts/provinces not contained in Sardegna region (indeed "Agrigento" is in "Sicilia", and not in "Sardegna").  

Thank you for your offer, but unfortunately I do not have the knowledge neither the time to join your great project as a developer. I am just a Master Degree student of Physics of Data in Padova that enjoys programming and can provide some humble suggestions, nothing more. Using your library for a lab project though! :)

**Comment by lisphilar** [comment]
With #561, dataset guide (docs/markdown/INSTALLATION.md) was updated! If you have any ideas, please let me know or create documentation issues/pull reuquests.
https://lisphilar.github.io/covid19-sir/markdown/INSTALLATION.html#recommended-datasets

Addition of city level data will be discussed in #562. It is difficult to reflect this idea at this time, but it is very interesting and useful for many users.

CovsirPhy project members are fully volunteers and we, including me, are not professors or datascience/software speciallists. Feel free to join our discussion via issues/pull requests...and advertise this project anywhere :P

**Comment by andrybicio** [comment]
Of course, I will try and give you some suggestions from my side! Thank you and to all developers!  :+1:
